### PR TITLE
[Utils] Add deprecate function and move testing_utils under utils

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# Improved deprecation functionality

We are introducing multiple deprecation cycles already and currently are missing structure on how to handle this exactly and by adding `**kwargs` to the scheduler config init we have lost functionality. More specifically:

- 1. We do not keep track when functionality should be removed as we forget to remove the deprecation cycle and the functionality (this happened twice now)
- 2. Right now when you pass a incorrect argument, e.g. `PNDMScheduler(num_inference_steps=5)` nothing happens because it's swallowed by **kwargs. **However** we should, just as before https://github.com/huggingface/diffusers/pull/651 throw a TypeError in this case.

=> This PR solves this by introducing a `deprecate` function (which should be mainly used by maintainers) that automatically forces us to remove deprecated functionality correctly + solves 2. by throwing a type error. In addition it removes boiler plate code and should make it more efficient for us to deprecate things.

The API of `deprecate` is as follows:

```python
new_arg = deprecate("old_arg", "0.7.0", "Please use `new_arg` instead", take_from=kwargs) or new_arg
```

=> This forces us to remove this statement when version >= 0.7.0, gives a nice default error message and makes sure that `kwargs` can only have deprecated args.

Please take a look into the changes here and the tests to better understand all the functionality

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 1]
```diff
@@ -59,15 +58,15 @@ def __init__(
...
                 f"The configuration file of this scheduler: {scheduler} is outdated. `steps_offset`"
                 f" should be set to 1 instead of {scheduler.config.steps_offset}. Please make sure "
                 "to update the config accordingly as leaving `steps_offset` might led to incorrect results"
                 " in future versions. If you have downloaded this checkpoint from the Hugging Face Hub,"
                 " it would be very nice if you could open a Pull request for the `scheduler/scheduler_config.json`"
-                " file",
-                DeprecationWarning,
+                " file"
             )
+            deprecate("steps_offset!=1", "1.0.0", deprecation_message, standard_warn=False)
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:69]: This is a big change and I like to keep it for a bit until we force people to not pass `steps_offset` anymore.

File: src/diffusers/schedulers/scheduling_ddim.py [hunk 2]
```diff
@@ -175,17 +174,10 @@ def set_timesteps(self, num_inference_steps: int, **kwargs):
...
-        if "offset" in kwargs:
-            warnings.warn(
-                "`offset` is deprecated as an input argument to `set_timesteps` and will be removed in v0.4.0."
-                " Please pass `steps_offset` to `__init__` instead.",
-                DeprecationWarning,
-            )
-
-            offset = kwargs["offset"]
+        deprecated_offset = deprecate(
+            "offset", "0.5.0", "Please pass `steps_offset` to `__init__` instead.", take_from=kwargs
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim.py:178]: bump as we would have already missed this version

File: src/diffusers/schedulers/scheduling_pndm.py [hunk 3]
```diff
@@ -155,16 +155,10 @@ def set_timesteps(self, num_inference_steps: int, **kwargs) -> torch.FloatTensor
...
-
-        if "offset" in kwargs:
-            warnings.warn(
-                "`offset` is deprecated as an input argument to `set_timesteps` and will be removed in v0.4.0."
-                " Please pass `steps_offset` to `__init__` instead."
-            )
-
-            offset = kwargs["offset"]
+        deprecated_offset = deprecate(
+            "offset", "0.5.0", "Please pass `steps_offset` to `__init__` instead.", take_from=kwargs
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_pndm.py:159]: bump as we would have already missed this version

File: src/diffusers/utils/outputs.py [hunk 4]
```diff
@@ -87,11 +87,7 @@ def __getitem__(self, k):
         if isinstance(k, str):
             inner_dict = {k: v for (k, v) in self.items()}
             if self.__class__.__name__ in ["StableDiffusionPipelineOutput", "ImagePipelineOutput"] and k == "sample":
-                warnings.warn(
-                    "The keyword 'samples' is deprecated and will be removed in version 0.4.0. Please use `.images` or"
-                    " `'images'` instead.",
-                    DeprecationWarning,
-                )
+                deprecate("samples", "0.6.0", "Please use `.images` or `'images'` instead.")
```
Comment (patrickvonplaten) [inline src/diffusers/utils/outputs.py:90]: bump twice as we would have already missed this version and it's important

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from diffusers import __version__
+from diffusers.utils import deprecate
+
+
+class DeprecateTester(unittest.TestCase):
```
Comment (patrickvonplaten) [inline tests/test_utils.py:22]: This test best describes all the functionality that we need to ensure

File: src/diffusers/utils/testing_utils.py [hunk 6]
```diff
@@ -22,6 +24,50 @@
...
+    if isinstance(deprecated_kwargs, dict) and len(deprecated_kwargs) > 0:
+        call_frame = inspect.getouterframes(inspect.currentframe())[1]
+        filename = call_frame.filename
+        line_number = call_frame.lineno
+        function = call_frame.function
+        key, value = next(iter(deprecated_kwargs.items()))
+        raise TypeError(f"{function} in {filename} line {line_number-1} got an unexpected keyword argument `{key}`")
+
+    if len(values) == 0:
+        return
```
Comment (patil-suraj) [inline src/diffusers/utils/testing_utils.py:65]: Very cool!

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for working on this, this will be very useful!
And really nice tests 😍 ! Love this way of explaining new features with tests.

File: src/diffusers/utils/testing_utils.py [hunk 7]
```diff
@@ -1,10 +1,12 @@
...
 import random
 import re
 import unittest
+import warnings
 from distutils.util import strtobool
 from pathlib import Path
-from typing import Union
+from typing import Any, Dict, Optional, Union
 
 import torch
```
Comment (pcuenca) [inline src/diffusers/utils/testing_utils.py:11]: I think this breaks `import diffusers` if PyTorch is not installed.

We should move `deprecate` to its own file. Not sure whether we should call it `deprecate.py`, `deprecation_utils.py` or something more general in case we need more functionality in the future.

**Review by pcuenca — COMMENTED** [review]
This is extremely cool, and so useful! It was very helpful to start by reading the tests.

Seeing that `deprecate` was inside `testing_utils.py` sounded a bit off, so I looked more closely to see what was there and realized that we would be breaking `import diffusers` when PyTorch is not installed. In any case, I believe that the semantics of the function call for moving it to another place.

File: src/diffusers/utils/testing_utils.py (hunk 7 above)
Comment (patrickvonplaten) [reply src/diffusers/utils/testing_utils.py:11]: Sounds good! I'll add it to `deprecation_utils.py` 

Regarding `import torch`, this should be done in another PR - this is unrelated to this PR (`import torch` was there before). We should definitely make sure though that the library doesn't break when `torch` is not installed.

Comment (pcuenca) [reply src/diffusers/utils/testing_utils.py]: > Regarding `import torch`, this should be done in another PR - this is unrelated to this PR (`import torch` was there before).

Yes, but `testing_utils` was only imported when testing, that's really the difference (not the `import`). I should have made it clearer in my comment :)
